File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.java
import java.util.*;

class Solution {
    public int solution(int[][] jobs) {
        int answer = 0;
        int end = 0;
        int jobsIdx = 0;
        int count = 0;
        
        Arrays.sort(jobs, (o1, o2) -> (o1[0] - o2[0]));
        PriorityQueue<int[]> pq = new PriorityQueue<>((o1, o2) -> o1[1] - o2[1]);
        
        while(count < jobs.length) {
            while(jobsIdx < jobs.length && jobs[jobsIdx][0] <= end){
                pq.add(jobs[jobsIdx++]);
            }
            
            if(pq.isEmpty()){
            end = jobs[jobsIdx][0]; // 다음 Job의 시작 시간으로 설정
            } else{
                int[] temp = pq.poll();
                answer += end + temp[1] - temp[0]; // 시작한 시간 + 소요 시간 - 하려고 했던 시간
                end += temp[1];
                count++;
            }
        }
        
        return (int) Math.floor(answer / jobs.length);
    }
}

File: 프로그래머스/3/42627. 디스크 컨트롤러/디스크 컨트롤러.py
from collections import deque
import heapq as hq

def solution(jobs):
    answer = 0 # 반환시각 - 요청시각
    time = 0
    heap = []
    index = 0
    
    jobs.sort()
    
    for i in range(len(jobs)):
        if jobs[i][0] == time:
            hq.heappush(heap, (jobs[i][1], jobs[i][0], i)) # 소요시간, 요청시각, 작업 번호
            index = i + 1
        else:
            break
    
    while index < len(jobs) or len(heap) > 0:
        if len(heap) > 0:
            clear_job = hq.heappop(heap)
            time += clear_job[0]
            # print("time:", time, "clear_job:", clear_job)
            answer += time - clear_job[1]

        # 현재 시간에 맞춰 들어온 작업 heap에 넣기
        while index < len(jobs):
            job = jobs[index]
            if(job[0] <= time):
                hq.heappush(heap, (job[1], job[0], index))
                index+=1
            else:
                break
        
        if len(heap) == 0:
            time += 1
                
    return answer // len(jobs)